packages/block-library/src/gallery/edit.js

```diff
@@ -71,146 +65,152 @@ const MOBILE_CONTROL_PROPS_RANGE_CONTROL = Platform.select( {
+
 	const { __unstableMarkNextChangeAsNotPersistent } = useDispatch(
 		'core/block-editor'
 	);
 
-	function setAttributes( newAttrs ) {
-		if ( newAttrs.ids ) {
-			throw new Error(
-				'The "ids" attribute should not be changed directly. It is managed automatically when "images" attribute changes'
-			);
-		}
+	const currentImageOptions = { linkTarget, linkTo, sizeSlug };
+	const [ imageSettings, setImageSettings ] = useState( currentImageOptions );
+	const [ dirtyImageOptions, setDirtyImageOptions ] = useState( false );
+	const [ images, setImages ] = useState( [] );
 
-		if ( newAttrs.images ) {
-			newAttrs = {
-				...newAttrs,
-				// Unlike images[ n ].id which is a string, always ensure the
-				// ids array contains numbers as per its attribute type.
-				ids: map( newAttrs.images, ( { id } ) => parseInt( id, 10 ) ),
-			};
+	useEffect( () => {
+		const currentOptionsState = ! isEqual(
+			currentImageOptions,
+			imageSettings
+		);
+		if ( currentOptionsState !== dirtyImageOptions ) {
+			setDirtyImageOptions( currentOptionsState );
 		}
+	}, [ currentImageOptions, imageSettings ] );
 
-		props.setAttributes( newAttrs );
-	}
-
-	function onSelectImage( index ) {
-		return () => {
-			setSelectedImage( index );
-		};
-	}
-
-	function onDeselectImage() {
-		return () => {
-			setSelectedImage();
-		};
-	}
-
-	function onMove( oldIndex, newIndex ) {
-		const newImages = [ ...images ];
-		newImages.splice( newIndex, 1, images[ oldIndex ] );
-		newImages.splice( oldIndex, 1, images[ newIndex ] );
-		setSelectedImage( newIndex );
-		setAttributes( { images: newImages } );
-	}
-
-	function onMoveForward( oldIndex ) {
-		return () => {
-			if ( oldIndex === images.length - 1 ) {
-				return;
-			}
-			onMove( oldIndex, oldIndex + 1 );
+	const { getBlock, getMedia, getSettings } = useSelect( ( select ) => {
+		return {
+			getBlock: select( 'core/block-editor' ).getBlock,
+			getSettings: select( 'core/block-editor' ).getSettings,
+			getMedia: select( 'core' ).getMedia,
 		};
-	}
+	}, [] );
 
-	function onMoveBackward( oldIndex ) {
-		return () => {
-			if ( oldIndex === 0 ) {
-				return;
-			}
-			onMove( oldIndex, oldIndex - 1 );
-		};
-	}
+	const innerBlockImages = useSelect( ( select ) => {
+		return select( 'core/block-editor' ).getBlock( clientId ).innerBlocks;
+	} );
 
-	function onRemoveImage( index ) {
-		return () => {
-			const newImages = filter( images, ( img, i ) => index !== i );
-			setSelectedImage();
-			setAttributes( {
-				images: newImages,
-				columns: attributes.columns
-					? Math.min( newImages.length, attributes.columns )
-					: attributes.columns,
-			} );
-		};
-	}
+	useEffect( () => {
+		const newImages = innerBlockImages.map( ( block ) => {
+			return {
+				id: block.attributes.id,
+				url: block.attributes.url,
+				attributes: block.attributes,
+				imageData: getMedia( block.attributes.id ),
+			};
+		} );
+		setImages( newImages );
+	}, [ innerBlockImages ] );
```

Comment:
Hey, WDYT about something like this? :)
```suggestion
	const images = useMemo(
		() =>
			innerBlockImages.map( ( block ) => ( {
				id: block.attributes.id,
				url: block.attributes.url,
				attributes: block.attributes,
				imageData: getMedia( block.attributes.id ),
			} ) ),
		[ innerBlockImages ]
	);
```

Then the state with images could be removed as well

Reply:
for sure, will definitely look at these optimisations now we have some of the basic flow issues sorted.

Reply:
Done